src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,237 @@
+
+    printf("Enumerating GC Heap Objects\n");
+    return pCorProfilerInfo->EnumerateGCHeapObjects(heap_walk_fn, &state);
+}
+
+HRESULT GCHeapEnumerationProfiler::ValidateEnumerateGCHeapObjects(HRESULT expected)
+{
+    HRESULT hr = EnumerateGCHeapObjects();
+
+    if (hr == expected)
+    {
+        if (FAILED(expected))
+        {
+            printf("Encountered exception as expected.\n");
+            _expectedExceptions++;
+            return S_OK;
+        }
+
+        printf("EnumerateGCHeapObjects succeeded.\n");
+        printf("Number of objects: %d\n", _objectsCount.load());
+        printf("Number of custom GCHeapObject types: %d\n", _customGCHeapObjectTypesCount.load());
+        return S_OK;
+    }
+
+    if (FAILED(hr) && FAILED(expected))
+    {
+        printf("EnumerateGCHeapObjects failed with exception hr=0x%x, but expected exception hr=0x%x\n", hr, expected);
+    }
+    else if (FAILED(expected))
+    {
+        printf("EnumerateGCHeapObjects succeeded, but expected exception hr=0x%x\n", expected);
+    }
+    else // FAILED(hr)
+    {
+        printf("Error: failed to enumerate GC heap objects. hr=0x%x\n", hr);
+    }
+
+    _failures++;
+    return E_FAIL;
+}
+
+extern "C" EXPORT void STDMETHODCALLTYPE EnumerateGCHeapObjects()
+{
+    printf("EnumerateGCHeapObjects PInvoke\n");
+    GCHeapEnumerationProfiler *instance = static_cast<GCHeapEnumerationProfiler*>(GCHeapEnumerationProfiler::Instance);
+    if (instance == nullptr)
+    {
+        printf("Error: profiler instance is null.\n");
+        return;
+    }
+    instance->ValidateEnumerateGCHeapObjects(0);
+}
+
+extern "C" EXPORT void STDMETHODCALLTYPE SuspendRuntime()
+{
+    printf("SuspendRuntime PInvoke\n");
+    GCHeapEnumerationProfiler *instance = static_cast<GCHeapEnumerationProfiler*>(GCHeapEnumerationProfiler::Instance);
+    if (instance == nullptr)
+    {
+        printf("Error: profiler instance is null.\n");
+        return;
+    }
+    instance->pCorProfilerInfo->SuspendRuntime();
+}
+
+extern "C" EXPORT void STDMETHODCALLTYPE ResumeRuntime()
+{
+    printf("ResumeRuntime PInvoke\n");
+    GCHeapEnumerationProfiler *instance = static_cast<GCHeapEnumerationProfiler*>(GCHeapEnumerationProfiler::Instance);
+    if (instance == nullptr)
+    {
+        printf("Error: profiler instance is null.\n");
+        return;
+    }
+    instance->pCorProfilerInfo->ResumeRuntime();
+}
+
+extern "C" EXPORT void STDMETHODCALLTYPE EnumerateHeapObjectsInBackgroundThread()
+{
+    GCHeapEnumerationProfiler* instance = static_cast<GCHeapEnumerationProfiler*>(GCHeapEnumerationProfiler::Instance);
+    if (instance == nullptr) {
+        printf("Error: profiler instance is null.\n");
+        return;
+    }
+
+    std::thread([instance]()
+                {
+                    std::this_thread::sleep_for(std::chrono::seconds(1));
+                    instance->ValidateEnumerateGCHeapObjects(0x80131388);
+                }).detach();
```

Comment:
If the test creates threads I'd keep track of them and make sure to join them later. Otherwise there is no guarantee the code being tested on these threads has ran prior to the test reporting its PASS/FAIL results. There is also the possibility that calling profiler API asynchronously during runtime shutdown/process exit could generate unexpected error conditions.

Reply:
Added the suggested logic at the start of Shutdown to wait for all threads tracked to finish execution with join().